softlearning/environments/adapters/dm_control_adapter.py

```diff
@@ -0,0 +1,155 @@
+    # generally, dm_control action_specs are of type BoundedArraySpec
+    # To handle dm_control observation_specs as inputs, we check the following input types in order
+    # to enable recursive calling on each ArraySpec nested item.
+    if isinstance(dm_control_space, BoundedArraySpec):
+        gym_box = spaces.Box(
+            low=dm_control_space.minimum,
+            high=dm_control_space.maximum,
+            shape=None,
+            dtype=dm_control_space.dtype)
+        # Note: `gym.Box` doesn't allow both shape and min/max to be defined
+        # at the same time. Thus we omit shape in the constructor and verify
+        # that it's been implicitly set correctly.
+        assert gym_box.shape == dm_control_space.shape, (
+            (gym_box.shape, dm_control_space.shape))
+        return gym_box
+    elif isinstance(dm_control_space, ArraySpec):
+        return spaces.Box(
+            low=-float("inf"),
+            high=float("inf"),
+            shape=dm_control_space.shape,
+            dtype=dm_control_space.dtype)
+    elif isinstance(dm_control_space, OrderedDict):
+        return spaces.Dict(OrderedDict([
+            (key, convert_dm_control_to_gym_space(value))
+            for key, value in dm_control_space.items()
+        ]))
+    else:
+        raise ValueError(dm_control_space)
+
+
+class DmControlAdapter(SoftlearningEnv):
+    """Adapter that implements the SoftlearningEnv for DeepMind Control Suite envs."""
+
+    def __init__(self,
+                 domain,
+                 task,
+                 *args,
+                 env=None,
+                 normalize=True,
+                 observation_keys=None,
+                 unwrap_time_limit=True,
+                 **kwargs):
+        assert not args, (
+            "Gym environments don't support args. Use kwargs instead.")
+
+        self.normalize = normalize
+        self.unwrap_time_limit = unwrap_time_limit
+
+        self._Serializable__initialize(locals())
+        super(DmControlAdapter, self).__init__(domain, task, *args, **kwargs)
+        if env is None:
+            assert (domain is not None and task is not None), (domain, task)
+            env = suite.load(
+                domain_name=domain,
+                task_name=task,
+                task_kwargs=kwargs
+                # TODO(hartikainen): Figure out how to pass kwargs to this guy.
+                # Need to split into `task_kwargs`, `environment_kwargs`, and `visualize_reward` bool.
+                # Check the suite.load(.) in:
+                # https://github.com/deepmind/dm_control/blob/master/dm_control/suite/__init__.py
+            )
+        else:
+            assert domain is None and task is None, (domain, task)
+
+        assert isinstance(env.observation_spec(), OrderedDict)
+        self.observation_keys = (
+            observation_keys or tuple(env.observation_spec().keys()))
+        # TODO(hartikainen): Do we need to normalize actions? For now, assume
```

Comment:
For now, solve this by checking the mix/max of action_space. If other than -1/1, raise `ValueError("Action space min/max should be -1/1")`.

Reply:
Yep, implemented.